universal-application-tool-0.0.1/app/assets/javascripts/azure_upload.ts

```diff
@@ -0,0 +1,83 @@
+/**
+ * This is responsible for uploading a file to Azure blob storage.
+ */
+
+class AzureUploadController {
+  static UPLOAD_CONTAINER_ID = 'azure-upload-form-component'
+
+  constructor() {
+    const uploadContainer = document.getElementById(
+      AzureUploadController.UPLOAD_CONTAINER_ID
+    )
+    uploadContainer.addEventListener('submit', (event) =>
+      this.attemptUpload(event, uploadContainer)
+    )
+  }
+
+  getValueFromInputLabel(label: string): string {
+    return (<HTMLInputElement>document.getElementsByName(label)[0]).value
+  }
+
+  attemptUpload(event: Event, uploadContainer: HTMLElement | null) {
+    event.preventDefault()
+    // This is set by an imported JavaScript Azure blob client.
+    // window["azblob"] was used because the TS compiler complained about using
+    // window.azblob.
+    const azblob = window['azblob']
+    if (uploadContainer == null) {
+      throw 'Attempted to upload to null container'
+    }
+    const azureUploadProps = this.getAzureUploadProps(uploadContainer)
+
+    const redirectUrl = new URL(azureUploadProps.successActionRedirect)
+
+    const blockBlobURL = azblob.BlockBlobURL.fromBlobURL(
+      new azblob.BlobURL(
+        `${azureUploadProps.blobUrl}?${azureUploadProps.sasToken}`,
```

Comment:
suggestion (blocking): could `azureUploadProps.blobUrl` as constructed on the server include the `sasToken` instead of doing string formatting here?

Reply:
We can do the string formatting in the BlobStorage class, but we have to do string formatting I believe because you have to get the blob URL and SAS token separately from the server. I could be wrong but based on my understanding.

Reply:
So, in the past, we were returning the full SAS URL from ```BlobStorage.java``` class, but made the decision to break that up because in all use cases except this one, we are using the blob URL separately from the SAS token. We can revert to the state where the BlobStorageClient is returning the full URL if you would prefer that, I'm just not sure it makes the most sense.

Reply:
Gotcha, fine to leave as-is then